server/lambda_function.py

```diff
@@ -0,0 +1,357 @@
+        logger.debug('Retrieving jwt token')
+        logger.debug('jwt token key: %s', key)
+        if 'cookie' not in map(lambda x: x.lower(), headers):
+            logger.info('no jwt token found for key %s', key)
+            return {}
+
+        cookie = cookies.SimpleCookie()
+        cookie.load(headers.get('cookie'))
+
+        if key in cookie:
+
+            # We take multiple secrets to allow for online secret rotation.
+            # The first secret is the current signer,
+            # and the others are potentially still in use, but will
+            # be rotated out.
+            for jwt_secret in jwt_secrets:
+                try:
+                    # github.com/jpadilla/pyjwt/blob/master/tests/test_api_jwt.py
+                    # This will validate:
+                    #   iss, aud
+                    # This will validate (if they are in the request):
+                    #   exp, nbf, iat
+                    return jwt.decode(
+                        cookie[key].value,
+                        jwt_secret,
+                        issuer=base_uri,
+                        audience=base_uri,
+                        algorithms=[algorithm]
+                    )
+                except jwt.exceptions.InvalidTokenError as e:
+                    # will catch exp, nbf, iat, iss, aud errors
+                    logger.error(
+                        'Exception decoding jwt token key: %s Exception: %s',
+                         key, e
+                    )
+                    continue
+
+        return {}
+
+    def with_user(event, context, callback):
+        logger.debug('Calling with_user')
+        """
+        Run the given request handler, passing it the user.
+        If the user has no valid auth token, then
+        redirect the user/resource owner to the keymaster
+        without invoking the given request handler.
+        """
+        jwt = get_jwt_token(event.get('headers'), oauth_token)
+        if oauth_token_key in jwt:
+            logger.info("User already auth'd")
+            user = dict(jwt)
+            for key in jwt_keys:
+                user.pop(key, None)
+            return callback(user)
+
+        logger.info("User not already auth'd")
+
+        client = OAuth2Session(
+            client_id,
+            scope="openid mail profile",
+            redirect_uri=redirect_uri
+        )
+        authorization_url, state = client.authorization_url(
+            authorization_endpoint)
+
+        # State is used to prevent CSRF, keep this for later.
+        cookie = Cookie()
+        initial_path = base_uri + event.get('path')
+        query = event.get('queryStringParameters')
+        if query:
+            initial_path += '?' + urllib.parse.urlencode(query)
+        logger.debug('initial_path: %s', initial_path)
+        state_token_params = jwt_params(
+            oauth_state=state,
+            initial_path=initial_path
+        )
+        cookie.add_jwt(oauth_state, state_token_params)
```

Comment:
There is no need for this cookie, you should piggyback your state with the jwt. See https://github.com/cviecco/cloud-gate/blob/with-local-openidc-handler/broker/httpd/auth.go

Reply:
I intended there to be only two cookies, one for the initial state with a shorter ttl, and the token with a longer ttl. Do you mean that one of these is unnecessary, or that there is another cookie?